src/serial.rs

```diff
@@ -0,0 +1,474 @@
+                    self.modem_status
+                }
+            }
+            SCR => self.scratch,
+            _ => 0,
+        }
+    }
+
+    /// Helps in sending more bytes to the guest in one shot, by storing
+    /// `input` bytes in UART buffer and letting the driver know there is
+    /// some pending data to be read by setting RDA bit and its corresponding
+    /// interrupt when not already triggered.
+    ///
+    /// # Arguments
+    /// * `input` - The data to be sent to the guest.
+    pub fn queue_raw_bytes(&mut self, input: &[u8]) -> Result<()> {
+        if !self.is_in_loop_mode() {
+            self.in_buffer.extend(input);
+            self.set_lsr_rda_bit();
+            self.received_data_interrupt()?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::sink;
+    use std::sync::{Arc, Mutex};
+
+    #[derive(Clone)]
+    struct SharedBuffer {
+        buf: Arc<Mutex<Vec<u8>>>,
+    }
+
+    impl SharedBuffer {
+        fn new() -> SharedBuffer {
+            SharedBuffer {
+                buf: Arc::new(Mutex::new(Vec::new())),
+            }
+        }
+    }
+
+    impl Write for SharedBuffer {
+        fn write(&mut self, buf: &[u8]) -> Result<usize> {
+            self.buf.lock().unwrap().write(buf)
+        }
+        fn flush(&mut self) -> Result<()> {
+            self.buf.lock().unwrap().flush()
+        }
+    }
+
+    static RAW_INPUT_BUF: [u8; 3] = [b'a', b'b', b'c'];
+
+    #[test]
+    fn test_serial_output() {
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let serial_out = SharedBuffer::new();
+        let mut serial = Serial::new(intr_evt, serial_out.clone());
+
+        // Valid one char at a time writes.
+        RAW_INPUT_BUF
+            .iter()
+            .for_each(|&c| serial.write(DATA, c).unwrap());
+        assert_eq!(serial_out.buf.lock().unwrap().as_slice(), &RAW_INPUT_BUF);
+    }
+
+    #[test]
+    fn test_serial_raw_input() {
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut serial = Serial::new(intr_evt.try_clone().unwrap(), sink());
+
+        serial.write(IER, IER_RDA_BIT).unwrap();
+        serial.queue_raw_bytes(&RAW_INPUT_BUF).unwrap();
+
+        // Verify the serial raised an interrupt.
+        assert_eq!(intr_evt.read().unwrap(), 1);
+
+        let mut lsr = serial.read(LSR);
+        assert_ne!(lsr & LSR_DATA_READY_BIT, 0);
+
+        // Verify reading the previously pushed buffer.
+        RAW_INPUT_BUF.iter().for_each(|&c| {
+            lsr = serial.read(LSR);
+            assert_ne!(lsr & LSR_DATA_READY_BIT, 0);
+            assert_eq!(serial.read(DATA), c);
+            assert_eq!(
+                serial.interrupt_identification,
+                DEFAULT_INTERRUPT_IDENTIFICATION
+            );
+        });
+
+        lsr = serial.read(LSR);
+        assert_eq!(lsr & LSR_DATA_READY_BIT, 0);
+
+        // Check if reading from the largest u8 offset returns 0.
```

Comment:
Why is this check useful?

Reply:
This was a copy paste test that kind of made sense, but it's not the most useful one.

I added a separate test, `test_invalid_access()` for "invalid" read/write accesses, but with offset `8`, hope it's better.